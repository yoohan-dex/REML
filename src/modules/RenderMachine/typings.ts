import { Tree, ChildrenType } from 'lib/lynx/lib/format';

export interface ResumeStore {
  staticComponents: StaticComponent[];
  globalData: StaticData[];
  globalTheme: StaticData[];
  liveComponents: LiveComponents[];
  liveData: LiveData[];
  messages: Messages;
  theme: Theme;
}

// export type ComponentType = 'Widget' | 'Layout';

export type StaticComponent = {
  // type: ComponentType;
  name: string;
  data: StaticData[];
  theme: StaticData[];
  children: Tree[] | ChildrenType;
  codebase: Code;
};

export type Code = {
  html: string;
  css: string;
};

export type Expression = {
  componentId: string;
  data: string[];
  expression: string;
};

export interface LiveComponents {
  id: string;
  children: LiveComponents[];
}

export type StaticData = {
  description: string;
  name: string;
  type: string;
  default: any;
  value: any;
};

export type LiveData = {
  id: string;
  data: StaticData[];
  theme: StaticData[];
};

export type Styles = {
  [key: string]: string;
};

export interface Message {
  name: string;
  description: string;
  type: string;
  default: any;
  value: any;
}

export type Messages = {
  [module: string]: {
    [name: string]: StaticData;
  }
};

export type Theme = {
  [module: string]: {
    [name: string]: StaticData;
  }
};
