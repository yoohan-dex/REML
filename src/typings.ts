declare module 'react-tap-event-plugin' {

  // tslint:disable-next-line:no-empty
  export function injectTapEventPlugin(): void;
}

declare module 'redux-responsive' {
  const responsiveStoreEnhancer: any;
  const responsiveStateReducer: any;
  export {
    responsiveStoreEnhancer,
    responsiveStateReducer,
  };
}
declare module 'neal-react' {
  const App: any;
  const Code: any;
  const CustomerQuote: any, CustomerQuotes: any;
  const DropdownMenu: any, DropdownToggle: any;
  const Footer: any;
  const Hero: any;
  const HorizontalSplit: any;
  const ImageList: any, ImageListItem: any;
  const Navbar: any, NavItem: any;
  const Page: any;
  const PricingPlan: any, PricingTable: any;
  const Section: any;
  const SignupInline: any, SignupModal: any;
  const Stripe: any;
  const Team: any;
  const TeamMember: any;
  export {
    App,
    Code,
    CustomerQuote, CustomerQuotes,
    DropdownMenu, DropdownToggle,
    Footer,
    Hero,
    HorizontalSplit,
    ImageList, ImageListItem,
    Navbar, NavItem,
    Page,
    PricingPlan, PricingTable,
    Section,
    SignupInline, SignupModal,
    Stripe,
    Team,
    TeamMember,
  };
}
declare module 'material-ui-responsive-drawer' {
  const ResponsiveDrawer: any;
  const BodyContainer: any;
  const ResponsiveAppBar: any;
  const responsiveDrawer: any;
  export {
    ResponsiveDrawer,
    BodyContainer,
    ResponsiveAppBar,
    responsiveDrawer,
  };
}

declare module 'react-device-simulator' {
  const DeviceSimulator: any;
  export default DeviceSimulator;
}

declare module 'react-frame-component' {
  const Frame: any;
  export default Frame;
}

declare module 'paul' {
  const Walk: any;
  export default Walk;
}

declare module 'redux-offline' {
  const offline: any;
  export { offline };
}

declare module 'redux-offline/lib/defaults' {
  const offlineConfig: any;
  export default offlineConfig;
}