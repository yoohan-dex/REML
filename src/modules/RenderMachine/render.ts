import * as React from 'react';
import * as R from 'ramda';
import { StaticData, StaticComponent, ResumeStore, Styles, LiveComponents, LiveData } from './typings';
import { ChildrenType, Text, Tree } from '../../lib/lynx/lib/format';
const paul = require('paul');
interface Scope {
  currentComponentName: string[];
  currentLiveComponentName: string[];
  currentComponentChildren: ChildrenType[];
  liveComponents: string[];
}

interface WalkOptions {
  key: number; // for react unique key
  scope: Scope; // for parse component
}

type Walk = (node: any, scope: LiveComponents[]) => any;

function parseExpression(str: string) {
  const leftPos = str.indexOf('{{');
  if (leftPos === -1) {
    return false;
  }
  const rightPos = str.indexOf('}}', leftPos);
  if (rightPos === -1) {
    return false;
  }
  const expression = str.slice(leftPos + 2, rightPos).trim();
  if (expression.includes('{') || expression.includes('}')) {
    throw (new Error('Expression can not includes any curly bracket'));
  }
  return {
    left: str.slice(0, leftPos),
    expression,
    right: str.slice(rightPos + 2),
  };
}

function renderText(
  node: Text,
  walk: Walk,
  store: ResumeStore,
  walkOptions: WalkOptions,
  scope: LiveComponents[],
): any {
  const expressionObject = parseExpression(node.content);
  if (!expressionObject) {
    return node.content;
  }
  const {
    left,
    expression,
    right,
  } = expressionObject;
  const rightObject: Text = {
    type: 'Text',
    content: right,
  };
  let data: StaticData;
  function extraExpression(): any {
    if (data.value) {
      return [left + data.value, renderText(rightObject, walk, store, walkOptions, scope)];
    } else if (data.default) {
      return [left + data.default, renderText(rightObject, walk, store, walkOptions, scope)];
    }
  }
  if (expression === 'children') {
    const rightSide = renderText(rightObject, walk, store, walkOptions, scope);
    if (walkOptions.scope.currentComponentName.length > 1) {
      // tslint:disable-next-line:no-unused-expression
      walkOptions.scope.currentComponentName.pop() as string;
    }
    const children = walk(walkOptions.scope.currentComponentChildren.pop(), scope);
    return [
      left,
      children,
      rightSide,
    ];
  } else if (expression.indexOf('::') !== -1) {
    const messages = expression.split('::');
    if (messages.length > 1) {
      const [module, name] = messages;
      data = store.messages[module][name];
      if (data) {
        return extraExpression();
      }
      throw new Error(`you have no the message item named ${name}`);
    }
  } else {
    data = getLiveOrStaticData(expression, walkOptions, store);
    if (data) {
      return extraExpression();
    } else {
      data = getData(expression, store.globalData);
      if (data) {
        return extraExpression();
      } else {
        throw new Error(`you have no a data named ${expression}`);
      }
    }
  }
}
export function getMessages(module: string, name: string, store: ResumeStore) {
  return store.messages[module][name];

}
export function getLiveOrStaticData(dataName: string, walkOptions: WalkOptions, store: ResumeStore) {
  let data: StaticData;
  const len = walkOptions.scope.currentLiveComponentName.length;
  if (len > 0) {

    const idx = R.findIndex(R.propEq('id', walkOptions.scope.currentLiveComponentName[len - 1]), store.liveData);
    data = getData(dataName, store.liveData[idx].data);
  } else {
    const currentComponent =
      getComponentFromStatic(
        walkOptions.scope.currentComponentName[walkOptions.scope.currentComponentName.length - 1],
        store,
      );
    data = getData(dataName, currentComponent.data);
  }
  return data;
}
export default function render(
  tree: Tree[] | ChildrenType,
  store: ResumeStore,
) {
  store.liveComponents = [];
  store.liveComponents[0] = {
    id: 'Resume',
    children: [],
  };
  let walkOptions: WalkOptions = {
    key: 0, // for the react component key
    scope: {
      currentComponentName: ['Resume'], // the component stack
      currentLiveComponentName: [],
      currentComponentChildren: [],
      liveComponents: ['Resume'],
    },  // for the component's inner data and theme;
  };

  return paul.walk(tree, (
    node: Tree | Text,
    walk: Walk,
    scope: LiveComponents[] = store.liveComponents[0].children,
    // what: any,
  ) => {
    if (node.type === 'Text') {
      return renderText(node, walk, store, walkOptions, scope);
    }

    const condition = parseCondition(node, store, walkOptions);
    const loop = parseLoop(node, store, walkOptions);
    if (condition) {
      if (condition.value || condition.default) {
        if (loop) {
          if (!Array.isArray(loop.default) || (loop.value && !Array.isArray(loop.value))) {
            throw (new Error(`${loop.name} should be an array`));
          }

          if (loop.value) {
            return loop.value.map((v: any) => renderElement(node, walk, store, walkOptions, scope));
          } else if (loop.default) {
            const components = loop.default.map((v: any) => {
              return renderElement(node, walk, store, walkOptions, scope);
            });
            return components;
          }
          // tslint:disable-next-line:no-console
          console.error('loop failed', `you have no the data named ${loop.name}`);
          return '';
        }
        return renderElement(node, walk, store, walkOptions, scope);

      }
      return '';

    } else if (loop) {
      if (!Array.isArray(loop.default) || (loop.value && !Array.isArray(loop.value))) {
        throw (new Error(`${loop.name} should be an array`));
      }

      if (loop.value) {
        return loop.value.map((v: any) => renderElement(node, walk, store, walkOptions, scope));
      } else if (loop.default) {
        return loop.default.map((v: any) =>
          renderElement(node, walk, store, walkOptions, scope),
        );
      }
      // tslint:disable-next-line:no-console
      console.error('loop failed', `you have no the data named ${loop.name}`);
      return '';
    } else {
      return renderElement(node, walk, store, walkOptions, scope);
    }
  });
}

function parseCondition(
  node: Tree,
  store: ResumeStore,
  walkOptions: WalkOptions
) {
  if (!node.attributes) {
    node.attributes = {};
  }
  if (node.attributes.condition) {
    const name = node.attributes.condition;
    let data = getLiveOrStaticData(name, walkOptions, store);
    if (!data) {
      data = getData(name, store.globalData);
      if (!data) {
        return ({
          name,
          value: '',
          default: '',
        });
      } else {
        return ({
          name,
          value: data.value,
          default: data.default,
        });
      }
    }
    return ({
      name,
      value: data.value,
      default: data.default,
    });

  }
  return false;
}

function getData(name: string, data: StaticData[]): StaticData {
  return R.find(
    R.propEq('name', name)
  )(data);
}
function parseLoop(
  node: Tree,
  store: ResumeStore,
  walkOptions: WalkOptions,
) {
  const {
    scope,
  } = walkOptions;
  if (node.attributes.loop) {
    const currentComponent = getComponentFromStatic(
      scope.currentComponentName[scope.currentComponentName.length - 1],
      store,
    );
    const name = node.attributes.loop;
    let data = getData(name, currentComponent.data);
    if (!data) {
      data = getData(name, store.globalData);
      if (!data) {
        return {
          name,
          value: '',
          default: [],
        };
      } else {
        return ({
          name,
          value: data.value,
          default: data.default,
        });
      }
    } else {
      return {
        name,
        value: data.value,
        default: data.default,
      };
    }
  }
  return false;
}

function getComponentFromStatic(currentComponentName: string, store: ResumeStore) {
  const component = R.find(
    R.propEq(
      'name',
      currentComponentName,
    ))(store.staticComponents);
  if (component) {
    return component as StaticComponent;
  }
  throw (new Error(`It is likely you have no the component named ${currentComponentName} in your components`));
}

function parseStyle(styles: Styles, store: ResumeStore, walkOptions: WalkOptions) {
  const inlineStyle: Styles = {};
  const { scope } = walkOptions;
  Object.keys(styles).forEach(key => {
    const isExpression = parseExpression(styles[key]);
    if (!isExpression) {
      inlineStyle[key] = styles[key];
    } else {
      const { left, expression, right } = isExpression;
      const currentComponent = getComponentFromStatic(
        scope.currentComponentName.length > 1
          ? scope.currentComponentName[scope.currentComponentName.length - 1]
          : scope.currentComponentName[0],
        store,
      );
      const data = getData(expression, currentComponent.data);
      inlineStyle[key] = left + data.value ? data.value : data.default + right;
    }
  });
  return inlineStyle;
}

function parseUrl(url: string, store: ResumeStore, walkOptions: WalkOptions) {
  let ret: string = '';
  const { scope } = walkOptions;
  console.log('parseURL', url);
  const isExpression = parseExpression(url);
  if (!isExpression) {
    ret = url;
  } else {
    const { left, expression, right } = isExpression;
    const currentComponent = getComponentFromStatic(
      scope.currentComponentName.length > 1
        ? scope.currentComponentName[scope.currentComponentName.length - 1]
        : scope.currentComponentName[0],
      store,
    );
    const data = getData(expression, currentComponent.data);
    if (data) {
      ret = left + (data.value ? data.value : data.default) + right;
    }

  }
  return ret;
}

function renderElement(
  node: Tree,
  walk: Walk,
  store: ResumeStore,
  walkOptions: WalkOptions,
  scope: LiveComponents[]) {
  const props: {
    [idx: string]: any;
  } = {
      key: walkOptions.key++,
    };
  if (node.attributes.styles) {
    props.style = { ...parseStyle(node.attributes.styles, store, walkOptions) };
  }
  if (node.attributes.className) {
    props.className = parseUrl(node.attributes.className.join(' '), store, walkOptions);
  }
  if (node.attributes.src) {
    props.src = parseUrl(node.attributes.src, store, walkOptions);
  }
  if (node.attributes.href) {
    props.href = parseUrl(node.attributes.href, store, walkOptions);
  }
  if (node.type === 'Element') {
    return React.createElement(
      node.tagName,
      props,
      walk(node.children, scope));
  } else if (node.type === 'Component') {
    const staticComponent = getComponentFromStatic(node.tagName, store);
    if (node.attributes.key) {
      const id = `${node.tagName}-${node.attributes.key}`;
      const liveComponent = {
        id,
        children: [],
      };
      let idx = -1;
      store.liveData.forEach((v, i) => {
        if (v.id === liveComponent.id) {
          idx = i;
        }
      });
      const liveData: LiveData = {
        id,
        data: staticComponent.data.slice(),
        theme: staticComponent.theme.slice(),
      };
      walkOptions.scope.currentLiveComponentName.push(id);
      if (idx === -1) {
        store.liveData.push(liveData);
      } else {
        liveData.data.forEach((d, i) => {
          if (store.liveData[idx].data[i]) {
            // ...
          } else {
            store.liveData[idx].data.push(d);
          }
        });
        liveData.theme.forEach((t, i) => {
          if (store.liveData[idx].theme[i]) {
            // ...
          } else {
            store.liveData[idx].theme.push(t);
          }
        });
      }
      scope.push(liveComponent);
    }
    walkOptions.scope.currentComponentName.push(node.tagName);
    walkOptions.scope.currentComponentChildren.push(node.children);
    const component = React.createElement(
      'div',
      {
        ...props,
        style: { display: 'inline' },
        className:
        node.attributes.key
          ? `${props.className} ${node.tagName}-${node.attributes.key}`
          : `${props.className} ${node.tagName}`,
        key: walkOptions.key++,
      },
      walk(staticComponent.children, node.attributes.key ? scope[scope.length - 1].children : scope),
    );
    if (node.attributes.key && walkOptions.scope.currentLiveComponentName.length > 0) {
      walkOptions.scope.currentLiveComponentName.pop();
    }
    if (walkOptions.scope.currentComponentName.length > 1) {
      walkOptions.scope.currentComponentName.pop();
    }
    return component;
  }
  return new Error('the node type is neither an Element nor a Component');
}
