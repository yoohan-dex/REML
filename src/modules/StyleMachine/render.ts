import { StaticData, StaticComponent, LiveData, Theme } from '../RenderMachine/typings';
import * as React from 'react';
import bulldozer from 'lib/bulldozer';

export default function render(staticComponent: StaticComponent[], liveData: LiveData[], globalTheme: Theme) {
  const components: React.ReactNode[] = [];
  staticComponent.forEach(v => {
    let css: string = '';
    try {
      css = bulldozer(v.codebase.css, v.name);
    } catch (e) {
      // ...
    }
    components.push(
      renderComponent(css, v.theme, v.name, globalTheme)
    );
  });
  liveData.forEach(d => {
    let css: string = '';
    let theme: StaticData[] = [];
    const [component] = d.id.split('-');
    staticComponent.forEach(c => {
      if (c.name === component) {
        css = bulldozer(c.codebase.css, d.id);
      }
    });
    liveData.forEach(data => {
      if (data.id === d.id) {
        theme = data.theme;
      }
    });
    components.push(
      renderComponent(css, theme, d.id, globalTheme),
    );
  });
  return components;
}

function renderComponent(css: string, data: StaticData[], key: string, globalTheme: Theme) {

  css = parseExpression(css, data, globalTheme);
  return React.createElement('style', { key }, css);
}

function parseExpression(str: string, data: StaticData[], globalTheme: Theme): string {
  const $position = str.indexOf('$');
  if ($position !== -1) {
    const beginPos = $position + 1;
    const $finalPos = str.indexOf('$', beginPos);
    let variable = '';
    if ($finalPos !== -1) {

      // const finalPos = parseVariable(str, beginPos);
      variable = str.slice(beginPos, $finalPos);
    }
    let value: string = '';
    if (variable) {
      if (variable.indexOf('::') !== -1) {
        const theme = variable.split('::');
        if (theme.length > 1) {
          const [module, name] = theme;
          const themedata = globalTheme[module][name];
          if (themedata) {
            if (themedata.value) {
              value = themedata.value;
            } else if (themedata.default) {
              value = themedata.default;
            }
          }
        }
      } else {
        data.forEach(v => {
          if (v.name === variable) {
            if (v.value) {
              value = v.value;
            } else if (v.default) {
              value = v.default;
            }
          }
        });
      }
    } else {
      return '';
    }
    return (
      str.slice(0, $position)
      + value
      + parseExpression(str.slice($finalPos + 1), data, globalTheme)
    );
  }
  return str;
}

function parseVariable(str: string, position: number) {
  if (isVariable(str[position])) {
    position += 1;
    parseVariable(str, position);
  }
  return position + 1;
}
function isVariable(char: string) {
  const code = char.charCodeAt(0);
  if ((code >= 97 && code <= 122) || (code >= 65 && code <= 90) || code === 95) {
    return true;
  }
  return false;
}