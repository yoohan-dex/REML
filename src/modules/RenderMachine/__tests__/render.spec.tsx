import * as React from 'react';
import 'jest';
import * as renderer from 'react-test-renderer';

import render from '../render';
import { ResumeStore } from '../typings';
import lynx from '../../../lib/lynx';

const globalStore: ResumeStore = {
  staticComponents: [{
    codebase: {
      html: '',
      css: '',
    },
    theme: [{
      description: '',
      name: 'primary',
      value: '',
      type: 'string',
      default: '#333',
    }],
    name: 'Hey',
    data: [{
      description: '',
      name: 'data',
      value: '',
      type: 'string',
      default: 'haha',
    }, {
      description: '',
      name: 'if',
      value: '',
      type: 'string',
      default: 'true',
    }],
    children: [],
  }, {
    codebase: {
      html: '',
      css: '',
    },
    theme: [{
      description: '',
      name: 'primary',
      value: '',
      type: 'string',
      default: '#333',
    }],
    name: 'Resume',
    data: [{
      description: '',
      name: 'if',
      value: '',
      type: 'string',
      default: 'true',
    }, {
      description: '',
      name: 'ifnot',
      value: '',
      type: 'string',
      default: '',
    }, {
      description: '',
      name: 'array',
      value: '',
      type: 'array',
      default: [1, 2, 3, 4],
    }],
    children: [],
  }, {
    codebase: {
      html: '',
      css: '',
    },
    theme: [{
      description: '',
      name: 'primary',
      value: '',
      type: 'string',
      default: '#333',
    }],
    name: 'HaveChildren',
    data: [{
      description: '',
      name: 'if',
      value: '',
      type: 'string',
      default: 'true',
    }, {
      description: '',
      name: 'ifnot',
      value: '',
      type: 'string',
      default: '',
    }, {
      description: '',
      name: 'array',
      value: '',
      type: 'array',
      default: [1, 2, 3, 4],
    }],
    children: [{
      type: 'Text',
      content: 'children: {{children}}'
    }],
  }],
  liveData: [],
  theme: {},
  globalData: [
    {
      description: '',
      name: 'fuck',
      value: '',
      type: 'string',
      default: 'haha',
    }, {
      description: '',
      name: 'hello',
      value: '',
      type: 'array',
      default: [1, 2, 3],
    }, {
      description: '',
      name: 'loop',
      value: '',
      type: 'array',
      default: ['1', '2'],
    }, {
      description: '',
      name: 'primary',
      value: '',
      type: 'string',
      default: '#333',
    }],
  globalTheme: [{
    description: '',
    name: 'primary',
    value: '',
    type: 'string',
    default: '#333',
  }],
  liveComponents: [{
    id: '',
    children: [],
  }],
  messages: {},
};
describe('common html test', () => {
  it('common test', () => {
    const html = `<div>
    <div>hello</div>
    </div>`;
    const data = lynx(html);
    const store = Object.assign({}, globalStore);
    const component = renderer.create(<div>{render(data, store)}</div>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should render a component', () => {
    const html = `<div>
    <Hey>hello</Hey>

    </div>`;
    const data = lynx(html);
    const store = Object.assign({}, globalStore);
    const component = renderer.create(<div>{render(data, store)}</div>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should mounted if true', () => {
    const html = `<div>
    <Hey  re:if={{if}}>hello</Hey>

    </div>`;
    const data = lynx(html);
    const store = Object.assign({}, globalStore);
    const component = renderer.create(<div>{render(data, store)}</div>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should disappear if not true', () => {
    const html = `<div re:if={{ifnot}}>
    <Hey>hello</Hey>

    </div>`;
    const data = lynx(html);
    const store = Object.assign({}, globalStore);
    const component = renderer.create(<div>{render(data, store)}</div>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should loop if there is a loop', () => {
    const html = `<div re:for={{array}}>
    <Hey>hello</Hey>

    </div>`;
    const data = lynx(html);
    const store = Object.assign({}, globalStore);
    const component = renderer.create(<div>{render(data, store)}</div>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should render a component with judage the condition', () => {
    const html = `<div>
    <Hey re:if={{array}}>hello</Hey>

    </div>`;
    const data = lynx(html);
    const store = Object.assign({}, globalStore);
    const component = renderer.create(<div>{render(data, store)}</div>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should render a component with a loop ', () => {
    const html = `<div>
    <Hey re:for={{array}}>hello</Hey>

    </div>`;
    const data = lynx(html);
    const store = Object.assign({}, globalStore);
    const component = renderer.create(<div>{render(data, store)}</div>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should render a component contained another component ', () => {
    const html = `<div>
      <HaveChildren re:for={{array}}>
        <Hey />
      </HaveChildren>
    </div>`;
    const data = lynx(html);
    const component = renderer.create(<div>{render(data, globalStore)}</div>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('data test', () => {
  it('should take the live data', () => {
    // ...
  });
});
