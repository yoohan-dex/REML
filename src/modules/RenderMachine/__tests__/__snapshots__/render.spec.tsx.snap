exports[`common html test common test 1`] = `
<div>
  <div>
    
        
    <div>
      hello
    </div>
    
        
  </div>
</div>
`;

exports[`common html test should disappear if not true 1`] = `
<div>
  
</div>
`;

exports[`common html test should loop if there is a loop 1`] = `
<div>
  <div>
    
        
    <div
      className="Hey" />
    
    
        
  </div>
  <div>
    
        
    <div
      className="Hey" />
    
    
        
  </div>
  <div>
    
        
    <div
      className="Hey" />
    
    
        
  </div>
  <div>
    
        
    <div
      className="Hey" />
    
    
        
  </div>
</div>
`;

exports[`common html test should mounted if true 1`] = `
<div>
  <div>
    
        
    <div
      className="Hey" />
    
    
        
  </div>
</div>
`;

exports[`common html test should render a component 1`] = `
<div>
  <div>
    
        
    <div
      className="Hey" />
    
    
        
  </div>
</div>
`;

exports[`common html test should render a component contained another component  1`] = `
<div>
  <div>
    
          
    <div
      className="HaveChildren">
      children: 
      
              
      <div
        className="Hey" />
      
            
      
    </div>
    <div
      className="HaveChildren">
      children: 
      
              
      <div
        className="Hey" />
      
            
      
    </div>
    <div
      className="HaveChildren">
      children: 
      
              
      <div
        className="Hey" />
      
            
      
    </div>
    <div
      className="HaveChildren">
      children: 
      
              
      <div
        className="Hey" />
      
            
      
    </div>
    
        
  </div>
</div>
`;

exports[`common html test should render a component with a loop  1`] = `
<div>
  <div>
    
        
    <div
      className="Hey" />
    <div
      className="Hey" />
    <div
      className="Hey" />
    <div
      className="Hey" />
    
    
        
  </div>
</div>
`;

exports[`common html test should render a component with judage the condition 1`] = `
<div>
  <div>
    
        
    <div
      className="Hey" />
    
    
        
  </div>
</div>
`;
